File: Lab/Tp1/PalindromoRecursivo.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

bool isEnd(char *str)
{
      return strcmp(str, "FIM\n") == 0;
}

bool isPalindromoRecursivo(char *str, int inicio, int fim)
{
      if (inicio >= fim)
      {
            printf("SIM\n");
            return true;
      }

      if (str[inicio] != str[fim])
      {
            printf("NAO\n");
            return false;
      }

      return isPalindromoRecursivo(str, inicio + 1, fim - 1);
}

bool isPalindromo(char *str)
{
      int length = strlen(str);
      return isPalindromoRecursivo(str, 0, length - 2);
}

int main()
{
      char str[1000];
      do
      {
            fgets(str, 1000, stdin);
            if (!isEnd(str))
            {
                  isPalindromo(str);
            }
      } while (!isEnd(str));
      return 0;
}

File: Lab/Tp1/PalindromoRecursivo.java
import java.util.Scanner;

public class PalindromoRecursivo {

      public static boolean isEnd(String str) {
            return str.equals("FIM");
      }

      public static boolean isPalindromo(String str) {
            if (str.length() <= 1) {
                  System.out.println("SIM");
                  return true;
            }
            if (str.charAt(0) != str.charAt(str.length() - 1)) {
                  System.out.println("NAO");
                  return false;
            }

            return isPalindromo(str.substring(1, str.length() - 1));
      }

      public static void main(String[] args) {
            Scanner scanner = new Scanner(System.in);
            String str;
            do {
                  str = scanner.nextLine();
                  if (!isEnd(str)) {
                        isPalindromo(str);
                  }
            } while (!isEnd(str));
            scanner.close();
      }
}
